### PR TITLE
[FEATURE] Add length normalized metrics for machine translation tasks

### LINKED ISSUE #1094: Add length normalized loss metrics in API
## Description
In typical machine translation tasks, training/validation metric loss is computed using some loss function normalized by the length of target sequence. For example, in `train_gnmt.py`, the metric is computed with the following code:

```python
loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
```
Current Mxnet `metric.loss` does not support length normalization. It will be great to add length normalized metric in the api.

## References
- https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L1661
- https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/train_gnmt.py

### PR DESCRIPTION
**Pull request by liuzh91**
## Description ##
- Add length normalized loss metric for machine translation task. 
- Test cases are included for the new metrics. 
- Remove Gluonnlp EvalMetric class

Fixes #1094

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

File: src/gluonnlp/metric/length_normalized_loss.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# specific language governing permissions and limitations
+# under the License.
+""" Length Normalized Loss """
+
+from mxnet import ndarray
+from .masked_accuracy import EvalMetric
+
+__all__ = ['LengthNormalizedLoss']
+
+class LengthNormalizedLoss(EvalMetric):
```
Comment (leezu) [inline src/gluonnlp/metric/length_normalized_loss.py]: Why use evaluation metric base class for a loss?

**Comment by leezu** [comment]
Please merge master due to https://github.com/dmlc/gluon-nlp/pull/1096

File: src/gluonnlp/metric/length_normalized_loss.py (hunk 1 above)
Comment (liuzh91) [reply src/gluonnlp/metric/length_normalized_loss.py]: `EvalMetric` is the metric base class. Every training/validation metric needs to inherit from the class, see https://github.com/apache/incubator-mxnet/blob/3e638b4fc4de959c7afc614facc336a796347c9d/python/mxnet/metric.py#L1661 Although it has `Eval` as the prefix, It does not directly related to validation. I have to admit the name is confusing.

BTW `MaskedAccuracy` also inherited from `EvalMetric` although it can only be used during training. See https://github.com/dmlc/gluon-nlp/blob/434187eb5e7bb898482021ffbc9c1648d0700608/src/gluonnlp/metric/masked_accuracy.py.

We may submit an issue to rename the class `EvalMetric`.

**Comment by mli** [comment]
Job PR-1095/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1095/4/index.html

File: src/gluonnlp/metric/length_normalized_loss.py (hunk 1 above)
Comment (leezu) [reply src/gluonnlp/metric/length_normalized_loss.py]: But here you are using some gluon-nlp EvalMetric and not the mxnet one `https://github.com/dmlc/gluon-nlp/blob/2db35f7e909e27d91b85abb8b959416688080296/src/gluonnlp/metric/masked_accuracy.py#L24`?

Comment (liuzh91) [reply src/gluonnlp/metric/length_normalized_loss.py]: the code of gluon-nlp `EvalMetric` is exactly the same as the mxnet one. I do not know why they copy and paste the `EvalMetric` code here. Because `MaskedAccuracy` inherit from this `EvalMetric`, I inherit my class from this one for consistency.

Comment (leezu) [reply src/gluonnlp/metric/length_normalized_loss.py]: Thanks for clarification. I looked at the PR that introduced the copy: https://github.com/dmlc/gluon-nlp/pull/503#discussion_r247326960

It's safe to remove the copy of `EvalMetric` in GluonNLP now and to replace it with the import from MXNet.

Comment (liuzh91) [reply src/gluonnlp/metric/length_normalized_loss.py]: OK, I will remove that

**Comment by mli** [comment]
Job PR-1095/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1095/5/index.html

**Comment by mli** [comment]
Job PR-1095/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1095/6/index.html

File: src/gluonnlp/metric/length_normalized_loss.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            output_names=output_names, label_names=label_names,
+            has_global_stats=True)
+
+    # Parameter labels should be a list in the form of  [target_sequence,
+    # target_seqauence_valid_length]
+    def update(self, labels, preds):
+        if not isinstance(labels, list) or len(labels) != 2:
+            raise ValueError('labels must be a list. Its first element should be'
+                             ' target sequence and the second element should be'
+                             'the valid length of sequence.')
```
Comment (leezu) [inline src/gluonnlp/metric/length_normalized_loss.py]: Should this be added to the docstring of the update function? (currently no docstring exists)

Comment (liuzh91) [reply src/gluonnlp/metric/length_normalized_loss.py]: Docstring added

**Comment by mli** [comment]
Job PR-1095/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1095/7/index.html

**Comment by mli** [comment]
Job PR-1095/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1095/8/index.html
